File: 68_binary-tree-postorder-traversal/binary-tree-postorder-traversal.java
/*
@Copyright:LintCode
@Author:   hanqiao
@Problem:  http://www.lintcode.com/problem/binary-tree-postorder-traversal
@Language: Java
@Datetime: 16-05-18 09:45
*/

/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */
public class Solution {
    /**
     * @param root: The root of binary tree.
     * @return: Postorder in ArrayList which contains node values.
     */
    public ArrayList<Integer> postorderTraversal(TreeNode root) {
        // write your code here
        ArrayList<Integer> result = new ArrayList<>();
        Stack<TreeNode> stack = new Stack<>();
        TreeNode prev = null;
        TreeNode curr = root;
        if (root == null) {
            return result;
        }
        
        stack.push(root);
        while (!stack.isEmpty()) {
           curr = stack.peek();
           if (prev == null || prev.left == curr || prev.right == curr) {
               if (curr.left != null) {
                   stack.push(curr.left);
               } else if (curr.right != null) {
                   stack.push(curr.right);
               }
           } else if (curr.left == prev) {
               if (curr.right != null) {
                   stack.push(curr.right);
               }
           } else {
               result.add(curr.val);
               stack.pop();
           }
           prev = curr;
        }
        return result;
    }
}

File: 68_binary-tree-postorder-traversal/binary-tree-postorder-traversal.py
# coding:utf-8
'''
@Copyright:LintCode
@Author:   hanqiao
@Problem:  http://www.lintcode.com/problem/binary-tree-postorder-traversal
@Language: Python
@Datetime: 16-05-18 10:27
'''

"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""


class Solution:
    """
    @param root: The root of binary tree.
    @return: Postorder in ArrayList which contains node values.
    """
    def postorderTraversal(self, root):
        # write your code here
        # iterative method
        result = []
        if root is None:
            return result
        stack = []
        prev = None
        curr = root
        stack.append(root)
        while len(stack) != 0:
            curr = stack[len(stack) - 1]
            if prev is None or prev.left == curr or prev.right == curr:
                if curr.left is not None:
                    stack.append(curr.left)
                elif curr.right is not None:
                    stack.append(curr.right)
            elif curr.left == prev:
                if curr.right is not None:
                    stack.append(curr.right)
            else:
                result.append(curr.val)
                stack.pop()
            prev = curr
        return result